lib/real-device-management.js

```diff
@@ -67,20 +68,51 @@ async function runRealDeviceReset(device, opts) {
  * @param {InstallOptions} [opts]
  */
 async function installToRealDevice(device, app, bundleId, opts) {
-  if (!device.udid || !app) {
-    log.debug('No device id or app, not installing to real device.');
+  if (!device.udid || !app || !bundleId) {
+    log.debug('No device id, app or bundle id, not installing to real device.');
     return;
   }
 
-  const {skipUninstall, strategy, timeout} = opts ?? {};
+  const {skipUninstall, strategy, timeout, shouldEnforceUninstall} = opts ?? {};
+  const shouldUninstallApp = !skipUninstall && bundleId;
 
-  if (!skipUninstall && bundleId && (await device.isAppInstalled(bundleId))) {
-    log.debug(`Reset requested. Removing app with id '${bundleId}' from the device`);
+  if (shouldUninstallApp) {
+    log.info(`Reset requested. Removing app with id '${bundleId}' from the device`);
     await device.remove(bundleId);
   }
   log.debug(`Installing '${app}' on device with UUID '${device.udid}'...`);
-  await device.install(app, timeout, strategy);
-  log.debug('The app has been installed successfully.');
+
+  try {
+    await device.install(app, timeout, strategy);
+    log.debug('The app has been installed successfully.');
+  } catch (e) {
+    // Do nothing if the session did not require to enforce uninstalling the application
```

Comment:
same as above. I would say it would be fine to always retry install by default and not have this condition
we could change this later if people complain about that, but altogether I try to always target positive scenarios where people actually want to have their app successfully installed rather than a test failure

Reply:
Okay, i just wanted to keep current behavior unless the request requires to enforce uninstallation explicitly. No other strong opinion I have. Then, I'll update title as well as feat

Reply:
removed in the latest code